x/ibc-transfer/types/trace.go

```diff
@@ -0,0 +1,196 @@
+	// validate correctness of port and channel identifiers
+	for i := 0; i < len(identifiers); i += 2 {
+		if err := host.PortIdentifierValidator(identifiers[i]); err != nil {
+			return sdkerrors.Wrapf(err, "invalid port ID at position %d", i)
+		}
+		if err := host.ChannelIdentifierValidator(identifiers[i+1]); err != nil {
+			return sdkerrors.Wrapf(err, "invalid channel ID at position %d", i)
+		}
+	}
+	return nil
+}
+
+// Validate performs a basic validation of the DenomTrace fields.
+func (dt DenomTrace) Validate() error {
+	// empty trace is accepted when token lives on the original chain
+	switch {
+	case dt.Trace == "" && dt.BaseDenom != "":
+		return nil
+	case strings.TrimSpace(dt.Trace) == "" && strings.TrimSpace(dt.BaseDenom) == "":
+		return fmt.Errorf("cannot have an empty trace and empty base denomination")
+	case dt.Trace != "" && strings.TrimSpace(dt.BaseDenom) == "":
+		return sdkerrors.Wrap(ErrInvalidDenomForTransfer, "denomination cannot be blank")
+	}
+
+	// NOTE: no base denomination validation
+
+	identifiers := strings.Split(dt.Trace, "/")
+	return validateTraceIdentifiers(identifiers)
+}
+
+// Traces defines a wrapper type for a slice of IdentifiedDenomTraces.
+type Traces []DenomTrace
+
+// Validate performs a basic validation of each denomination trace info.
+func (t Traces) Validate() error {
+	seenTraces := make(map[string]bool)
+	for i, trace := range t {
+		hash := trace.Hash().String()
+		if seenTraces[hash] {
+			return fmt.Errorf("duplicated denomination trace with hash  %s", trace.Hash())
+		}
+
+		if err := trace.Validate(); err != nil {
+			return sdkerrors.Wrapf(err, "failed denom trace %d validation", i)
+		}
+		seenTraces[hash] = true
+	}
+	return nil
+}
+
+// ValidatePrefixedDenom checks that the denomination for an IBC fungible token packet denom is correctly prefixed.
+// The function will return no error if the given string follows one of the two formats:
+//
+//  - Prefixed denomination: '{portIDN}/{channelIDN}/.../{portID0}/{channelID0}/baseDenom'
+//  - Unprefixed denomination: 'baseDenom'
+func ValidatePrefixedDenom(denom string) error {
+	denomSplit := strings.Split(denom, "/")
+	if denomSplit[0] == denom && strings.TrimSpace(denom) != "" {
+		// NOTE: no base denomination validation
+		return nil
+	}
+
+	if strings.TrimSpace(denomSplit[len(denomSplit)-1]) == "" {
+		return sdkerrors.Wrap(ErrInvalidDenomForTransfer, "base denomination cannot be blank")
+	}
+
+	identifiers := denomSplit[:len(denomSplit)-1]
+	return validateTraceIdentifiers(identifiers)
+}
+
+// ValidateIBCDenom validates that the given denomination is either:
+//
+//  - A valid base denomination (eg: 'uatom')
+//  - A valid trace prefixed denomination  (eg: '{portIDN}/{channelIDN}/.../{portID0}/{channelID0}/baseDenom')
+//  - A valid fungible token representation (i.e 'ibc/{hash}') per ADR 001 https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-001-coin-source-tracing.md
+func ValidateIBCDenom(denom string) error {
```

Comment:
As mentioned above, it is confusing to allow both options in the state machine itself. Let's make things consistent and standardized in the state-machine, and allow either format at the CLI/REST level